{
 "metadata": {
  "language": "Julia",
  "name": "",
  "signature": "sha256:bc034ad304756eb065c0404f7a3589ca4cead2cbb7bd97644480ea398abce423"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "; cat ../COSim.jl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "module COSim\n",
        "using Distributions\n",
        "tempPos=Array(Float64,100000)\n",
        "tempOri=Array(Int64,  100000)\n",
        "\n",
        "# Types and Methods for Information on the Genome\n",
        "\n",
        "type LocusInfo\n",
        "    map_pos::Float64\n",
        "    allele_freq::Array\n",
        "end\n",
        "\n",
        "function get_num_alleles(my)\n",
        "    return length(my.allele_freq)\n",
        "end\n",
        "\n",
        "type ChromosomeInfo\n",
        "    chrLength::Float64\n",
        "    numLoci::Int64\n",
        "    mapPos::Array{Float64,1}\n",
        "    loci::Array{LocusInfo,1}\n",
        "end\n",
        "\n",
        "function set_num_loci(my,n::Int64)\n",
        "    my.numLoci=n;\n",
        "end\n",
        "\n",
        "function get_num_loci(my)\n",
        "    return my.numLoci\n",
        "end\n",
        "\n",
        "function mk_mappos_from_locus_info(my)\n",
        "    my.mapPos.resize(my.chrLength)\n",
        "    for i in 1:my.chrLength\n",
        "        my.mapPos[i]=my.locus[i].map_pos\n",
        "    end\n",
        "end\n",
        "\n",
        "type GenomeInfo\n",
        "    chr::Array{ChromosomeInfo,1}\n",
        "    numChrom::Int64\n",
        "    mutRate::Float64\n",
        "end\n",
        "\n",
        "function set_num_chrom(my,n::Int64)\n",
        "    my.numChrom=n; nothing\n",
        "end\n",
        "\n",
        "function get_num_chrom(my)\n",
        "    return my.numChrom\n",
        "end\n",
        "\n",
        "function make_map_pos(my)\n",
        "    for i in 1:my.numChrom\n",
        "        mk_mappos_from_locus_info(my.chr[i])\n",
        "    end\n",
        "    mapPosDone = true\n",
        "end\n",
        "\n",
        "#Types and Methods for Simulating Genotypes of Animals\n",
        "\n",
        "type Chromosome\n",
        "    haplotype::Array{Int64,1}\n",
        "    ori::Array{Int64,1}\n",
        "    pos::Array{Float64,1}\n",
        "end\n",
        "\n",
        "type Animal\n",
        "    genomePat::Array{Chromosome,1}\n",
        "    genomeMat::Array{Chromosome,1}\n",
        "    myID::Int64\n",
        "    sireID::Int64\n",
        "    damID::Int64\n",
        "end\n",
        "\n",
        "function Animal(mySire::Int64, myDam::Int64)\n",
        "    my=Animal(Array(Chromosome,common.G.numChrom),Array(Chromosome,common.G.numChrom),0,0,0)\n",
        "    my.sireID = mySire\n",
        "    my.damID  = myDam\n",
        "    my.myID   = common.countId\n",
        "    common.countId += 1\n",
        "    for i in 1:common.G.numChrom\n",
        "        my.genomePat[i]=Chromosome(Array(Int64,0),Array(Int64,0),Array(Float64,0))\n",
        "        my.genomeMat[i]=Chromosome(Array(Int64,0),Array(Int64,0),Array(Float64,0))\n",
        "    end\n",
        "    return my\n",
        "end\n",
        "\n",
        "type Cohort\n",
        "    animalCohort::Array{Animal,1}\n",
        "    npMatrix::Array{Int64,2}\n",
        "end\n",
        "\n",
        "# Type for Storing Globals\n",
        "\n",
        "type CommonToAnimals\n",
        "    founders::Array{Animal,1}\n",
        "    G::GenomeInfo\n",
        "    countChromosome::Int64\n",
        "    countId::Int64\n",
        "end\n",
        "\n",
        "function sampleFounder()\n",
        "    my = Animal(0,0)\n",
        "    initFounderPosOri(my)\n",
        "    initFounderHaps(my)\n",
        "    return(my)\n",
        "end\n",
        "\n",
        "function initFounderHaps(my)\n",
        "\n",
        "    numberChromosomePair=get_num_chrom(common.G)\n",
        "\n",
        "    for i in 1:numberChromosomePair\n",
        "\n",
        "        numLoci=common.G.chr[i].numLoci\n",
        "\n",
        "        Base.resize!(my.genomePat[i].haplotype,numLoci)\n",
        "        Base.resize!(my.genomeMat[i].haplotype,numLoci)\n",
        "\n",
        "        for j in 1:numLoci\n",
        "            p=common.G.chr[i].loci[j].allele_freq[1]\n",
        "            my.genomePat[i].haplotype[j]=rand(Bernoulli(p))\n",
        "            my.genomeMat[i].haplotype[j]=rand(Bernoulli(p))\n",
        "        end\n",
        "    end\n",
        "end\n",
        "\n",
        "function initFounderPosOri(my)\n",
        "        numberChromosomePair=get_num_chrom(common.G)\n",
        "        for i in 1:numberChromosomePair\n",
        "            my.genomePat[i].ori=[common.countChromosome]\n",
        "            my.genomePat[i].pos=[0.0]\n",
        "\n",
        "            my.genomeMat[i].ori=[common.countChromosome+1]\n",
        "            my.genomeMat[i].pos=[0.0]\n",
        "        end\n",
        "        common.countChromosome += 2\n",
        "end\n",
        "\n",
        "function sampleFounders(numAnimals::Int64)\n",
        "        my=Cohort(Array(Animal,0),Array(Int64,0,0))\n",
        "\n",
        "        println(\"Sampling \",numAnimals,\" animals into base population.\")\n",
        "        resize!(my.animalCohort,numAnimals)\n",
        "        for i in 1:numAnimals\n",
        "            animal=sampleFounder()\n",
        "            my.animalCohort[i] = animal\n",
        "            push!(common.founders,animal)\n",
        "        end\n",
        "        return(my)\n",
        "end\n",
        "\n",
        "function getRandomInd(my::Cohort)\n",
        "    cohortSize=length(my.animalCohort)\n",
        "    thisOne=rand(1:cohortSize)\n",
        "    return(my.animalCohort[thisOne])\n",
        "end\n",
        "\n",
        "function sampleChildren(fathers::Cohort,mothers::Cohort,numAnimals::Int64)\n",
        "    my=Cohort(Array(Animal,0),Array(Int64,0,0))\n",
        "\n",
        "    #println(\"Sampling \",numAnimals,\" animals into next generation.\")\n",
        "    resize!(my.animalCohort,numAnimals)\n",
        "\n",
        "    for i in 1:numAnimals\n",
        "        animal=sampleNonFounder(getRandomInd(fathers),getRandomInd(mothers))\n",
        "        #println(\"Sampled animal: \",animal.myID)\n",
        "        my.animalCohort[i]=animal\n",
        "    end\n",
        "\n",
        "    return(my)\n",
        "end\n",
        "\n",
        "function sampleNonFounder(father,mother)\n",
        "        my=Animal(father.myID,mother.myID)\n",
        "\n",
        "        numberChromosomePair=get_num_chrom(common.G)\n",
        "        resize!(my.genomePat,numberChromosomePair)\n",
        "        resize!(my.genomeMat,numberChromosomePair)\n",
        "\n",
        "        sampleMyPosOri(my,father,mother)\n",
        "        return(my)\n",
        "end\n",
        "\n",
        "function sampleMyPosOri(my,father,mother)\n",
        "  sampleOnePosOri(my.genomePat,father)\n",
        "  sampleOnePosOri(my.genomeMat,mother)\n",
        " #=\n",
        "    a = @spawn sampleOnePosOri(my.genomePat,father)\n",
        "    b = @spawn sampleOnePosOri(my.genomeMat,mother)\n",
        "    wait(a)\n",
        "    wait(b)\n",
        "=#\n",
        "end\n",
        "\n",
        "function getMeanPosSize(my::Animal)\n",
        "    numberChromosomePair=get_num_chrom(common.G)\n",
        "    sum = 0.0\n",
        "    for i in 1:numberChromosomePair\n",
        "      sum += (length(my.genomePat[i].pos) + length(my.genomePat[i].pos))/2.0\n",
        "    end\n",
        "  return sum/numberChromosomePair\n",
        "end\n",
        "\n",
        "function sampleOnePosOri(genome::Array{Chromosome,1},parent::Animal)\n",
        "    numberChromosomePair=get_num_chrom(common.G)\n",
        "\n",
        "    for i in 1:numberChromosomePair\n",
        "\n",
        "        genome[i]=Chromosome(Array(Int64,0),Array(Int64,1),Array(Float64,1))\n",
        "\n",
        "        currentChrom=(rand(Bernoulli(0.5))==1)?parent.genomePat[i]:parent.genomeMat[i]\n",
        "\n",
        "        chrLength=common.G.chr[i].chrLength\n",
        "\n",
        "        binomialN=convert(Int64,ceil(chrLength*3+1))\n",
        "        numCrossover=rand(Binomial(binomialN,chrLength/binomialN))\n",
        "        rec=[0.0]\n",
        "\n",
        "        for irec in 1:numCrossover\n",
        "            push!(rec,chrLength*rand(1)[1])\n",
        "        end\n",
        "\n",
        "        push!(rec,chrLength)\n",
        "        sort!(rec)                #rec is like 0.00,0.23,0.45,0.76,1.00\n",
        "\n",
        "        numTemp=1\n",
        "\n",
        "        for j in 1:(length(rec)-1)\n",
        "            for k in 1:length(currentChrom.pos)\n",
        "              if currentChrom.pos[k] >= rec[j] && currentChrom.pos[k] < rec[j+1]\n",
        "                tempPos[numTemp]=currentChrom.pos[k]\n",
        "                tempOri[numTemp]=currentChrom.ori[k]\n",
        "                numTemp=numTemp+1\n",
        "              elseif currentChrom.pos[k]>=rec[j+1]\n",
        "              \tbreak\n",
        "              end\n",
        "            end\n",
        "\n",
        "\n",
        "            currentChrom=(currentChrom==parent.genomeMat[i])?parent.genomePat[i]:parent.genomeMat[i]\n",
        "\n",
        "            findRecOri=0\n",
        "            m=1\n",
        "            lengthChrPos = length(currentChrom.pos)\n",
        "            while m<=lengthChrPos && currentChrom.pos[m]<=rec[j+1] #pos[m] cannot be the length of chromosome\n",
        "               m += 1\n",
        "               findRecOri += 1\n",
        "            end\n",
        "            tempPos[numTemp]=rec[j+1]\n",
        "            tempOri[numTemp]=currentChrom.ori[findRecOri]\n",
        "            numTemp=numTemp+1 #**            \t\n",
        "        end\n",
        "\n",
        "        numTemp=numTemp-1 #remove the last one got from **\n",
        "        numTemp=numTemp-1 #remove the last one which is the length of chromsome\n",
        "\n",
        "        resize!(genome[i].pos,numTemp)\n",
        "        resize!(genome[i].ori,numTemp)\n",
        "\n",
        "        ##merging of nearby same ori;not sure whether more effcient\n",
        "\n",
        "        genome[i].pos[1]=tempPos[1]\n",
        "        genome[i].ori[1]=tempOri[1]\n",
        "        this=1\n",
        "        for(m in 2:numTemp)\n",
        "            if(tempOri[m]!=tempOri[m-1])\n",
        "                this=this+1\n",
        "                genome[i].pos[this]=tempPos[m]\n",
        "                genome[i].ori[this]=tempOri[m]\n",
        "            end\n",
        "        end\n",
        "\n",
        "        resize!(genome[i].pos,this) #resize after merging\n",
        "        resize!(genome[i].ori,this)\n",
        "\n",
        "    end\n",
        "end\n",
        "\n",
        "function getMyHaps(my)\n",
        "    numberChromosomePair=get_num_chrom(common.G)\n",
        "\n",
        "    for i in 1:numberChromosomePair\n",
        "        numLoci=common.G.chr[i].numLoci\n",
        "        resize!(my.genomePat[i].haplotype,numLoci)\n",
        "        resize!(my.genomeMat[i].haplotype,numLoci)\n",
        "\n",
        "\n",
        "        numOriPat=length(my.genomePat[i].ori)\n",
        "        push!(my.genomePat[i].pos,common.G.chr[i].chrLength) #this may make it not efficient\n",
        "\n",
        "        for segment in 1:numOriPat\n",
        "            whichFounder=iceil(my.genomePat[i].ori[segment]/2)\n",
        "            genomePatorMatInThisFounder=(my.genomePat[i].ori[segment]%2==0)?common.founders[whichFounder].genomeMat[i]:common.founders[whichFounder].genomePat[i]\n",
        "            choose = my.genomePat[i].pos[segment] .<= common.G.chr[i].mapPos .< my.genomePat[i].pos[segment+1]\n",
        "            my.genomePat[i].haplotype[choose]=genomePatorMatInThisFounder.haplotype[choose]\n",
        "        end\n",
        "\n",
        "        pop!(my.genomePat[i].pos)\n",
        "\n",
        "\n",
        "\n",
        "        numOriMat=length(my.genomeMat[i].ori)\n",
        "        push!(my.genomeMat[i].pos,common.G.chr[i].chrLength)#this may make it not efficient\n",
        "\n",
        "\n",
        "        for segment in 1:numOriMat\n",
        "            whichFounder=iceil(my.genomeMat[i].ori[segment]/2)\n",
        "            genomePatorMatInThisFounder=(my.genomeMat[i].ori[segment]%2==0)?common.founders[whichFounder].genomeMat[i]:common.founders[whichFounder].genomePat[i]\n",
        "            choose = my.genomeMat[i].pos[segment] .<= common.G.chr[i].mapPos .< my.genomeMat[i].pos[segment+1]\n",
        "            my.genomeMat[i].haplotype[choose]=genomePatorMatInThisFounder.haplotype[choose]\n",
        "        end\n",
        "\n",
        "        pop!(my.genomeMat[i].pos)\n",
        "\n",
        "    end\n",
        "end\n",
        "\n",
        "function printMyHaps(my)\n",
        "    numberChromosomePair=get_num_chrom(common.G)\n",
        "    println(\"I'm an animal with ID( \", my.myID,\" ) and sire( \", my.sireID, \" ) and dam( \", my.damID,\" )\")\n",
        "\n",
        "    for i in 1:numberChromosomePair\n",
        "        println(my.genomePat[i].haplotype)\n",
        "        println(my.genomePat[i].pos)\n",
        "        println(my.genomeMat[i].haplotype)\n",
        "        println(my.genomeMat[i].pos)\n",
        "    end\n",
        "end\n",
        "\n",
        "\n",
        "function getOurHaps(my::Cohort)\n",
        "    for i in my.animalCohort\n",
        "        getMyHaps(i)\n",
        "    end\n",
        "end\n",
        "\n",
        "function getOurGenotypes(my::Cohort)\n",
        "    getOurHaps(my)\n",
        "    npMatrix=Array(Int64,length(my.animalCohort), common.G.numChrom*common.G.chr[1].numLoci)\n",
        "    for (i,value) in enumerate(my.animalCohort)\n",
        "        npMatrix[i,:]=getMyGenotype(value)\n",
        "    end\n",
        "    return npMatrix\n",
        "end\n",
        "\n",
        "function getMyGenotype(my)\n",
        "    myGenotype=Array(Int64,0)\n",
        "    for i in 1:common.G.numChrom\n",
        "        append!(myGenotype, my.genomePat[i].haplotype+my.genomeMat[i].haplotype)\n",
        "    end\n",
        "    #println(myGenotype)\n",
        "    return myGenotype\n",
        "end\n",
        "\n",
        "function printOurHaps(my::Cohort)\n",
        "    for i in my.animalCohort\n",
        "        printMyHaps(i)\n",
        "    end\n",
        "end\n",
        "\n",
        "function getMeanPosSize(my::Cohort)\n",
        "  sum = 0.0\n",
        "    for i in my.animalCohort\n",
        "        sum += getMeanPosSize(i)\n",
        "    end\n",
        "  sum/length(my.animalCohort)\n",
        "end\n",
        "\n",
        "\n",
        "# Make object for storing globals\n",
        "\n",
        "G = GenomeInfo(Array(ChromosomeInfo,0),0,0.0)\n",
        "\n",
        "common = CommonToAnimals(Array(Animal,0),G,0,0)\n",
        "\n",
        "type JSimMembers\n",
        "    popSample::Function\n",
        "    popNew::Function\n",
        "    getGenotypes::Function\n",
        "    parents::Cohort\n",
        "    children::Cohort\n",
        "    gen::Int64\n",
        "end\n",
        "\n",
        "function popSampleW(numGen::Int64,popSize::Int64,my::JSimMembers)\n",
        "    if size(my.parents.animalCohort,1)==0\n",
        "        my.children = sampleFounders(popSize)\n",
        "        my.parents  = my.children\n",
        "        numGen -=1\n",
        "        my.gen += 1\n",
        "    end\n",
        "    for i in 1:numGen\n",
        "        my.gen += 1\n",
        "        println(\"Sampling \",popSize,\" animals into generation: \",my.gen)\n",
        "        my.children = sampleChildren(my.parents,my.parents,popSize)\n",
        "        my.parents  = my.children\n",
        "        #println(\"Mean Pos Size \",getMeanPosSize(my.parents))\n",
        "    end\n",
        "end\n",
        "\n",
        "function popSampleW(popSize::Int64,ancestors::JSimMembers)\n",
        "    newPop = COSim.startPop()\n",
        "    newPop.parents = ancestors.children\n",
        "    newPop.popSample(1,popSize)\n",
        "    return newPop\n",
        "end\n",
        "\n",
        "function popCross(popSize::Int64,breed1::JSimMembers,breed2::JSimMembers)\n",
        "    newPop = COSim.startPop()\n",
        "    newPop.children = sampleChildren(breed1.children,breed2.children,popSize)\n",
        "    return newPop\n",
        "end\n",
        "\n",
        "function getGenotypesW(my)\n",
        "    getOurGenotypes(my.children)\n",
        "end\n",
        "\n",
        "function init(numChr::Int64,numLoci::Int64,chrLength::Float64, geneFreq::Array{Float64,1},\n",
        "        mapPos::Array{Float64,1},mutRate::Float64,myCommon=common)\n",
        "\n",
        "    #create genome\n",
        "    locus_array = Array(LocusInfo,numLoci)\n",
        "    for i in 1:numLoci\n",
        "        locus_array[i] = LocusInfo(mapPos[i],[geneFreq[i],1-geneFreq[i]])\n",
        "    end\n",
        "    chromosome = ChromosomeInfo(chrLength,numLoci,mapPos,locus_array)\n",
        "    chr = fill(chromosome,numChr)\n",
        "    G = GenomeInfo(chr,numChr,mutRate)\n",
        "\n",
        "    # Init common\n",
        "\n",
        "    myCommon.founders=Array(Animal,0)\n",
        "    myCommon.G = G\n",
        "    myCommon.countId = 1\n",
        "    myCommon.countChromosome = 1; nothing\n",
        "end\n",
        "\n",
        "function startPop()\n",
        "\n",
        "    parents  = Cohort(Array(Animal,0),Array(Int64,0,0))\n",
        "    children = Cohort(Array(Animal,0),Array(Int64,0,0))\n",
        "\n",
        "    function popSample(numGen::Int64,popSize::Int64)\n",
        "        popSampleW(numGen,popSize,members)\n",
        "    end\n",
        "    function popNew(popSize::Int64)\n",
        "        popSampleW(popSize,members)\n",
        "    end\n",
        "    function getGenotypes()\n",
        "        getGenotypesW(members)\n",
        "    end\n",
        "    members = JSimMembers(popSample,popNew,getGenotypes,parents,children,0)\n",
        "    return(members)\n",
        "end\n",
        "end # of module COSim\n"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "include (\"../COSim.jl\")\n",
      "using COSim"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "chrLength = 1.0\n",
      "numChr    = 2\n",
      "numLoci   = 5\n",
      "mutRate   = 0.0\n",
      "locusInt  = chrLength/numLoci\n",
      "mapPos   = [0:locusInt:(chrLength-0.0001)]\n",
      "geneFreq = fill(0.5,numLoci)\n",
      "\n",
      "COSim.init(numChr,numLoci,chrLength,geneFreq,mapPos,mutRate) # init COSim"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "jsim1 = COSim.startPop() # init jsim1\n",
      "jsim2 = COSim.startPop() # init jsim2\n",
      "nothing"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ngen    = 3\n",
      "popSize = 10\n",
      "jsim1.popSample(ngen,popSize)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Sampling 10"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " animals into base population.\n",
        "Sampling 10 animals into generation: 2\n",
        "Sampling "
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "10 animals into generation: 3\n"
       ]
      }
     ],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ngen    = 3\n",
      "popSize = 10\n",
      "jsim2.popSample(ngen,popSize)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Sampling 10"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " animals into base population.\n",
        "Sampling 10 animals into generation: 2\n",
        "Sampling 10 animals into generation: 3\n"
       ]
      }
     ],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "jsim3 = jsim1.popNew(10) # start jsim3 by sampling from jsim1\n",
      "nothing"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Sampling 10"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " animals into generation: 1\n"
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "jsim4 = jsim1.popNew(10) # start jsim4 by sampling from jsim1\n",
      "nothing"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Sampling 10"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " animals into generation: 1\n"
       ]
      }
     ],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "jsim5 = COSim.popCross(30,jsim1,jsim2) # start jsim5 by crossing jsim1 and jsim2\n",
      "nothing"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "M = jsim5.getGenotypes()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 9,
       "text": [
        "30x10 Array{Int64,2}:\n",
        " 0  0  2  2  1  1  0  2  1  0\n",
        " 2  0  1  1  2  2  2  2  2  2\n",
        " 2  1  1  1  1  2  1  1  0  1\n",
        " 1  0  1  0  1  2  2  2  0  0\n",
        " 2  1  1  2  1  0  2  2  1  1\n",
        " 2  1  1  2  2  2  1  1  1  1\n",
        " 1  0  1  1  2  2  2  2  0  1\n",
        " 0  0  1  0  1  1  2  2  0  1\n",
        " 1  0  2  2  2  2  1  2  1  1\n",
        " 1  2  0  1  1  2  1  2  2  2\n",
        " 2  0  0  0  1  2  0  1  1  1\n",
        " 1  0  1  1  2  1  2  1  2  2\n",
        " 1  0  2  1  2  1  2  2  0  1\n",
        " \u22ee              \u22ee            \n",
        " 1  0  2  0  1  2  2  2  1  2\n",
        " 1  0  2  0  2  1  2  2  2  2\n",
        " 2  2  1  2  1  2  0  1  2  2\n",
        " 1  0  2  0  1  1  1  2  0  1\n",
        " 2  0  1  1  2  2  2  2  0  1\n",
        " 2  1  0  1  1  1  1  2  1  2\n",
        " 1  0  2  1  2  2  1  2  0  1\n",
        " 0  0  2  0  1  2  2  2  2  2\n",
        " 1  1  1  0  2  1  1  2  2  1\n",
        " 2  0  1  1  1  0  1  1  1  1\n",
        " 2  1  2  0  1  1  2  1  1  2\n",
        " 0  1  1  1  1  2  1  2  2  2"
       ]
      }
     ],
     "prompt_number": 9
    }
   ],
   "metadata": {}
  }
 ]
}