{
 "metadata": {
  "language": "Julia",
  "name": "",
  "signature": "sha256:55ac0f228092e2fc8fca6d64fad91b18d262f9e5d7130815cb072ec2d9cda25e"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "module COSim\n",
      "\n",
      "using Distributions\n",
      "\n",
      "tempPos=Array(Float64,100000)\n",
      "tempOri=Array(Int64,  100000)\n",
      "\n",
      "# Types and Methods for Information on the Genome\n",
      "\n",
      "type LocusInfo\n",
      "    map_pos::Float64\n",
      "    allele_freq::Array\n",
      "end\n",
      "\n",
      "function get_num_alleles(my)\n",
      "    return length(my.allele_freq)\n",
      "end\n",
      "\n",
      "type ChromosomeInfo\n",
      "    chrLength::Float64\n",
      "    numLoci::Int64\n",
      "    mapPos::Array{Float64,1}\n",
      "    loci::Array{LocusInfo,1}\n",
      "end\n",
      "\n",
      "function set_num_loci(my,n::Int64)\n",
      "    my.numLoci=n; nothing\n",
      "end\n",
      "\n",
      "function get_num_loci(my)\n",
      "    return my.numLoci\n",
      "end\n",
      "\n",
      "function mk_mappos_from_locus_info(my)\n",
      "    my.mapPos.resize(my.chrLength)\n",
      "    for i in 1:my.chrLength\n",
      "        my.mapPos[i]=my.locus[i].map_pos\n",
      "    end\n",
      "end\n",
      "\n",
      "type GenomeInfo\n",
      "    chr::Array{ChromosomeInfo,1}\n",
      "    numChrom::Int64\n",
      "    mutRate::Float64\n",
      "end\n",
      "\n",
      "function set_num_chrom(my,n::Int64)\n",
      "    my.numChrom=n; nothing\n",
      "end\n",
      "\n",
      "function get_num_chrom(my)\n",
      "    return my.numChrom\n",
      "end\n",
      "\n",
      "function make_map_pos(my)\n",
      "    for i in 1:my.numChrom\n",
      "        mk_mappos_from_locus_info(my.chr[i])\n",
      "    end\n",
      "    mapPosDone = true\n",
      "end\n",
      "\n",
      "#Types and Methods for Simulating Genotypes of Animals\n",
      "\n",
      "type Chromosome\n",
      "    haplotype::Array{Int64,1}\n",
      "    ori::Array{Int64,1}\n",
      "    pos::Array{Float64,1}\n",
      "end\n",
      "\n",
      "type Animal\n",
      "    genomePat::Array{Chromosome,1}\n",
      "    genomeMat::Array{Chromosome,1}\n",
      "    myID::Int64\n",
      "    sireID::Int64\n",
      "    damID::Int64\n",
      "end\n",
      "\n",
      "function Animal(mySire::Int64, myDam::Int64)\n",
      "    my=Animal(Array(Chromosome,common.G.numChrom),Array(Chromosome,common.G.numChrom),0,0,0)\n",
      "    my.sireID = mySire\n",
      "    my.damID  = myDam\n",
      "    my.myID   = common.countId\n",
      "    common.countId += 1\n",
      "    for i in 1:common.G.numChrom\n",
      "        my.genomePat[i]=Chromosome(Array(Int64,0),Array(Int64,0),Array(Float64,0))\n",
      "        my.genomeMat[i]=Chromosome(Array(Int64,0),Array(Int64,0),Array(Float64,0))\n",
      "    end\n",
      "    return my\n",
      "end\n",
      "\n",
      "type Cohort\n",
      "    animalCohort::Array{Animal,1}\n",
      "    npMatrix::Array{Int64,2}\n",
      "end\n",
      "\n",
      "# Type for Storing Globals\n",
      "\n",
      "type CommonToAnimals\n",
      "    founders::Array{Animal,1}\n",
      "    G::GenomeInfo\n",
      "    countChromosome::Int64\n",
      "    countId::Int64\n",
      "end\n",
      "\n",
      "function sampleFounder()\n",
      "    my = Animal(0,0)\n",
      "    initFounderPosOri(my)\n",
      "    initFounderHaps(my)\n",
      "    return(my)\n",
      "end\n",
      "\n",
      "function initFounderHaps(my)\n",
      "   \n",
      "    numberChromosomePair=get_num_chrom(common.G)\n",
      "    \n",
      "    for i in 1:numberChromosomePair\n",
      "        \n",
      "        numLoci=common.G.chr[i].numLoci\n",
      "        \n",
      "        Base.resize!(my.genomePat[i].haplotype,numLoci)\n",
      "        Base.resize!(my.genomeMat[i].haplotype,numLoci)\n",
      "\n",
      "        for j in 1:numLoci\n",
      "            p=common.G.chr[i].loci[j].allele_freq[1]\n",
      "            my.genomePat[i].haplotype[j]=rand(Bernoulli(p))\n",
      "            my.genomeMat[i].haplotype[j]=rand(Bernoulli(p))\n",
      "        end\n",
      "    end   \n",
      "end    \n",
      "\n",
      "function initFounderPosOri(my)\n",
      "        numberChromosomePair=get_num_chrom(common.G)\n",
      "        for i in 1:numberChromosomePair\n",
      "            my.genomePat[i].ori=[common.countChromosome]\n",
      "            my.genomePat[i].pos=[0.0]\n",
      "\n",
      "            my.genomeMat[i].ori=[common.countChromosome+1]\n",
      "            my.genomeMat[i].pos=[0.0]\n",
      "        end\n",
      "        common.countChromosome += 2\n",
      "end\n",
      "\n",
      "function sampleFounders(numAnimals::Int64)\n",
      "        my=Cohort(Array(Animal,0),Array(Int64,0,0))\n",
      "\n",
      "        println(\"Sampling \",numAnimals,\" animals into base population.\")\n",
      "        resize!(my.animalCohort,numAnimals)\n",
      "        for i in 1:numAnimals\n",
      "        animal=sampleFounder()\n",
      "            my.animalCohort[i] = animal\n",
      "            push!(common.founders,animal)\n",
      "        end\n",
      "        return(my)\n",
      "end\n",
      "\n",
      "function getRandomInd(my::Cohort)\n",
      "    cohortSize=length(my.animalCohort)\n",
      "    thisOne=rand(1:cohortSize)\n",
      "    return(my.animalCohort[thisOne])\n",
      "end\n",
      "\n",
      "function sampleChildren(fathers::Cohort,mothers::Cohort,numAnimals::Int64)\n",
      "    my=Cohort(Array(Animal,0),Array(Int64,0,0))\n",
      "    resize!(my.animalCohort,numAnimals)\n",
      "\n",
      "    for i in 1:numAnimals\n",
      "        animal=sampleNonFounder(getRandomInd(fathers),getRandomInd(mothers))\n",
      "        #println(\"Sampled animal: \",animal.myID)\n",
      "        my.animalCohort[i]=animal\n",
      "    end\n",
      "    \n",
      "    return(my)\n",
      "end\n",
      "\n",
      "function sampleNonFounder(father,mother)\n",
      "        my=Animal(father.myID,mother.myID)\n",
      "\n",
      "        numberChromosomePair=get_num_chrom(common.G)\n",
      "        resize!(my.genomePat,numberChromosomePair)\n",
      "        resize!(my.genomeMat,numberChromosomePair)\n",
      "\n",
      "        sampleMyPosOri(my,father,mother)\n",
      "        return(my)\n",
      "end\n",
      "\n",
      "function sampleMyPosOri(my,father,mother)\n",
      "    sampleOnePosOri(my.genomePat,father)\n",
      "    sampleOnePosOri(my.genomeMat,mother)\n",
      "end\n",
      "\n",
      "function sampleOnePosOri(genome::Array{Chromosome,1},parent::Animal)\n",
      "    numberChromosomePair=get_num_chrom(common.G)\n",
      "\n",
      "    for i in 1:numberChromosomePair\n",
      "\n",
      "        genome[i]=Chromosome(Array(Int64,0),Array(Int64,1),Array(Float64,1))\n",
      "\n",
      "        currentChrom=(rand(Bernoulli(0.5))==1)?parent.genomePat[i]:parent.genomeMat[i]\n",
      "\n",
      "        chrLength=common.G.chr[i].chrLength\n",
      "\n",
      "        binomialN=convert(Int64,ceil(chrLength*3+1))\n",
      "        numCrossover=rand(Binomial(binomialN,chrLength/binomialN))\n",
      "        rec=[0.0]\n",
      "\n",
      "        for irec in 1:numCrossover\n",
      "            push!(rec,chrLength*rand(1)[1])\n",
      "        end\n",
      "\n",
      "        push!(rec,chrLength)\n",
      "        sort!(rec)                #rec is like 0.00,0.23,0.45,0.76,1.00\n",
      "\n",
      "        numTemp=1\n",
      "        \n",
      "        for j in 1:(length(rec)-1)\n",
      "            \n",
      "            for k in 1:length(currentChrom.pos)\n",
      "              if currentChrom.pos[k] >= rec[j] && currentChrom.pos[k] < rec[j+1]\n",
      "                tempPos[numTemp]=currentChrom.pos[k]\n",
      "                tempOri[numTemp]=currentChrom.ori[k]\n",
      "                numTemp=numTemp+1\n",
      "              elseif currentChrom.pos[k]>=rec[j+1]\n",
      "                break\n",
      "              end\n",
      "            end\n",
      "            \n",
      "            currentChrom=(currentChrom==parent.genomeMat[i])?parent.genomePat[i]:parent.genomeMat[i]\n",
      "\n",
      "            findRecOri=0\n",
      "            m=1\n",
      "            lengthChrPos = length(currentChrom.pos)\n",
      "            while m<=lengthChrPos && currentChrom.pos[m]<=rec[j+1] #pos[m] cannot be the length of chromosome\n",
      "               m += 1\n",
      "               findRecOri += 1\n",
      "            end            \n",
      "            tempPos[numTemp]=rec[j+1]\n",
      "            tempOri[numTemp]=currentChrom.ori[findRecOri]\n",
      "            numTemp=numTemp+1 #**\n",
      "        end\n",
      "        \n",
      "        numTemp=numTemp-1 #remove the last one got from **\n",
      "        numTemp=numTemp-1 #remove the last one which is the length of chromsome\n",
      "\n",
      "        resize!(genome[i].pos,numTemp) \n",
      "        resize!(genome[i].ori,numTemp)\n",
      "        \n",
      "        ##merging of nearby same ori;not sure whether more effcient\n",
      "        \n",
      "        genome[i].pos[1]=tempPos[1]\n",
      "        genome[i].ori[1]=tempOri[1]\n",
      "        this=1\n",
      "        for(m in 2:numTemp)\n",
      "            if(tempOri[m]!=tempOri[m-1])\n",
      "                this=this+1\n",
      "                genome[i].pos[this]=tempPos[m]\n",
      "                genome[i].ori[this]=tempOri[m]\n",
      "            end\n",
      "        end\n",
      " \n",
      "        resize!(genome[i].pos,this) #resize after merging\n",
      "        resize!(genome[i].ori,this)\n",
      "        \n",
      "        #println(genome[i].pos)\n",
      "        #println(genome[i].ori)\n",
      "\n",
      "    end\n",
      "end\n",
      "  \n",
      "function getMyHaps(my)\n",
      "    getOneHaps(my.genomePat)\n",
      "    getOneHaps(my.genomeMat)\n",
      "end\n",
      "\n",
      "function getOneHaps(genome::Array{Chromosome,1})\n",
      "    numberChromosomePair=get_num_chrom(common.G)\n",
      "    \n",
      "    for i in 1:numberChromosomePair\n",
      "        numLoci=common.G.chr[i].numLoci\n",
      "        resize!(genome[i].haplotype,numLoci)\n",
      "\n",
      "        numOri=length(genome[i].ori)\n",
      "        push!(genome[i].pos,common.G.chr[i].chrLength) #this may make it not efficient\n",
      "        \n",
      "        iLoci   = 1\n",
      "        position = common.G.chr[i].mapPos[iLoci]\n",
      "        \n",
      "        for segment in 1:numOri\n",
      "            whichFounder=iceil(genome[i].ori[segment]/2)\n",
      "            genomePatorMatInThisFounder=(genome[i].ori[segment]%2==0)?common.founders[whichFounder].genomeMat[i]:common.founders[whichFounder].genomePat[i]\n",
      "            \n",
      "            startPos = genome[i].pos[segment]\n",
      "            endPos   = genome[i].pos[segment+1]\n",
      "            segLen   = 0\n",
      "                                        \n",
      "            while position >= startPos && position<endPos\n",
      "                iLoci = iLoci+1\n",
      "                segLen = segLen+1\n",
      "\n",
      "                if iLoci>numLoci\n",
      "                    break\n",
      "                end\n",
      "                \n",
      "                position = common.G.chr[i].mapPos[iLoci]\n",
      "            end\n",
      "                                   \n",
      "            endLoci=iLoci-1\n",
      "            genome[i].haplotype[(endLoci-segLen+1):endLoci]=genomePatorMatInThisFounder.haplotype[(endLoci-segLen+1):endLoci]\n",
      "        end\n",
      "        \n",
      "        pop!(genome[i].pos)\n",
      "    end\n",
      "end\n",
      "\n",
      "function printMyHaps(my)\n",
      "    numberChromosomePair=get_num_chrom(common.G)\n",
      "    println(\"I'm an animal with ID( \", my.myID,\" ) and sire( \", my.sireID, \" ) and dam( \", my.damID,\" )\")\n",
      "\n",
      "    for i in 1:numberChromosomePair\n",
      "        println(my.genomePat[i].haplotype)\n",
      "        println(my.genomePat[i].pos)\n",
      "        println(my.genomeMat[i].haplotype)\n",
      "        println(my.genomeMat[i].pos)\n",
      "    end\n",
      "end\n",
      "\n",
      "\n",
      "function getOurHaps(my::Cohort)\n",
      "    for i in my.animalCohort\n",
      "        getMyHaps(i)\n",
      "    end\n",
      "end\n",
      "\n",
      "function getOurGenotypes(my::Cohort)\n",
      "    getOurHaps(my)\n",
      "    npMatrix=Array(Int64,length(my.animalCohort), common.G.numChrom*common.G.chr[1].numLoci)\n",
      "    for (i,value) in enumerate(my.animalCohort)\n",
      "        npMatrix[i,:]=getMyGenotype(value)\n",
      "    end\n",
      "    return npMatrix\n",
      "end\n",
      "\n",
      "function getMyGenotype(my)\n",
      "    myGenotype=Array(Int64,0)\n",
      "    for i in 1:common.G.numChrom\n",
      "        append!(myGenotype, my.genomePat[i].haplotype+my.genomeMat[i].haplotype)\n",
      "    end\n",
      "    #println(myGenotype)\n",
      "    return myGenotype\n",
      "end\n",
      "\n",
      "function printOurHaps(my::Cohort)\n",
      "    for i in my.animalCohort\n",
      "        printMyHaps(i)\n",
      "    end\n",
      "end\n",
      "\n",
      "\n",
      "# Make object for storing globals\n",
      "\n",
      "G = GenomeInfo(Array(ChromosomeInfo,0),0,0.0)\n",
      "\n",
      "common = CommonToAnimals(Array(Animal,0),G,0,0)\n",
      "\n",
      "type JSimMembers\n",
      "    popSample::Function\n",
      "    popNew::Function\n",
      "    getGenotypes::Function\n",
      "    parents::Cohort\n",
      "    children::Cohort\n",
      "    gen::Int64\n",
      "end\n",
      "\n",
      "function popSampleW(numGen::Int64,popSize::Int64,my::JSimMembers)\n",
      "    if size(my.parents.animalCohort,1)==0\n",
      "        my.children = sampleFounders(popSize)\n",
      "        my.parents  = my.children\n",
      "        numGen -=1\n",
      "    end\n",
      "    for i in 1:numGen\n",
      "        my.gen += 1\n",
      "        println(\"Sampling \",popSize,\" animals into generation: \",my.gen)\n",
      "        my.children = sampleChildren(my.parents,my.parents,popSize)\n",
      "        my.parents  = my.children\n",
      "    end\n",
      "end\n",
      "\n",
      "function popSampleW(popSize::Int64,ancestors::JSimMembers)\n",
      "    newPop = COSim.startPop()\n",
      "    newPop.parents = ancestors.children\n",
      "    newPop.popSample(1,popSize)\n",
      "    return newPop\n",
      "end\n",
      "\n",
      "function popCross(popSize::Int64,breed1::JSimMembers,breed2::JSimMembers)\n",
      "    newPop = COSim.startPop()\n",
      "    newPop.children = sampleChildren(breed1.children,breed2.children,popSize)\n",
      "    return newPop\n",
      "end\n",
      "\n",
      "function getGenotypesW(my)\n",
      "    getOurGenotypes(my.children)\n",
      "end\n",
      "\n",
      "function init(numChr::Int64,numLoci::Int64,chrLength::Float64,geneFreq::Array{Float64,1},\n",
      "        mapPos::Array{Float64,1},mutRate::Float64,myCommon=common)       \n",
      "\n",
      "    #create genome\n",
      "    locus_array = Array(LocusInfo,numLoci)\n",
      "    for i in 1:numLoci\n",
      "        locus_array[i] = LocusInfo(mapPos[i],[geneFreq[i],1-geneFreq[i]])\n",
      "    end\n",
      "    chromosome = ChromosomeInfo(chrLength,numLoci,mapPos,locus_array)\n",
      "    chr = fill(chromosome,numChr)\n",
      "    G = GenomeInfo(chr,numChr,mutRate)\n",
      "\n",
      "    # Init common\n",
      "\n",
      "    myCommon.founders=Array(Animal,0)\n",
      "    myCommon.G = G\n",
      "    myCommon.countId = 1\n",
      "    myCommon.countChromosome = 1; nothing\n",
      "end\n",
      "\n",
      "function startPop()\n",
      "    \n",
      "    parents  = Cohort(Array(Animal,0),Array(Int64,0,0))\n",
      "    children = Cohort(Array(Animal,0),Array(Int64,0,0))\n",
      "    \n",
      "    function popSample(numGen::Int64,popSize::Int64)\n",
      "        popSampleW(numGen,popSize,members)\n",
      "    end\n",
      "    function popNew(popSize::Int64)\n",
      "        popSampleW(popSize,members)\n",
      "    end\n",
      "    function getGenotypes()\n",
      "        getGenotypesW(members)\n",
      "    end\n",
      "    members = JSimMembers(popSample,popNew,getGenotypes,parents,children,0)\n",
      "    return(members)\n",
      "end\n",
      "\n",
      "end # of COSimModule"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "using COSim\n",
      "\n",
      "chrLength = 1.0\n",
      "numChr    = 1\n",
      "numLoci   = 100\n",
      "mutRate   = 0.0\n",
      "\n",
      "locusInt  = chrLength/numLoci\n",
      "mapPos   = [0:locusInt:(chrLength-0.0001)]\n",
      "geneFreq = fill(0.5,numLoci)\n",
      "\n",
      "\n",
      "COSim.init(numChr,numLoci,chrLength,geneFreq,mapPos,mutRate)\n",
      "\n",
      "jsim1 = COSim.startPop()\n",
      "M=jsim1.getGenotypes()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 2,
       "text": [
        "0x100 Array{Int64,2}"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ngen    = 10\n",
      "popSize = 10\n",
      "jsim1.popSample(ngen,popSize)\n",
      "M=jsim1.getGenotypes()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Sampling 10"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " animals into base population.\n",
        "Sampling 10 animals into generation: 1\n",
        "Sampling 10 animals into generation: 2\n",
        "Sampling 10 animals into generation: 3\n",
        "Sampling 10 animals into generation: 4\n",
        "Sampling 10 animals into generation: 5\n",
        "Sampling 10 animals into generation: 6\n",
        "Sampling 10 animals into generation: 7\n",
        "Sampling 10 animals into generation: 8\n",
        "Sampling 10 animals into generation: 9\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 3,
       "text": [
        "10x100 Array{Int64,2}:\n",
        " 1  0  2  2  1  1  1  2  0  2  1  1  1  \u2026  1  2  1  1  2  0  1  1  1  0  2  0\n",
        " 1  0  2  2  1  0  2  1  1  1  0  1  2     0  2  0  2  2  0  2  2  2  0  2  2\n",
        " 0  0  0  2  0  2  0  2  0  2  2  0  0     1  2  0  2  2  0  2  0  0  0  2  0\n",
        " 1  0  1  2  1  1  1  1  1  1  1  1  1     2  2  2  0  2  0  0  1  1  0  2  2\n",
        " 0  0  2  2  0  2  0  2  0  2  2  0  0     2  2  2  0  2  0  0  0  0  0  2  2\n",
        " 1  0  2  2  1  1  1  1  1  1  1  1  1  \u2026  1  2  2  0  2  0  1  1  1  0  2  2\n",
        " 0  0  1  2  0  2  0  2  0  2  2  0  0     2  2  2  0  2  0  0  0  0  0  2  2\n",
        " 2  0  2  2  2  0  2  1  1  1  0  1  1     0  2  0  2  2  0  2  1  1  0  2  1\n",
        " 1  0  1  2  0  2  0  2  0  2  2  0  0     1  2  1  1  2  0  1  1  1  0  2  2\n",
        " 0  0  1  2  0  2  0  2  0  2  2  0  0     2  2  2  0  2  0  0  0  0  0  2  2"
       ]
      }
     ],
     "prompt_number": 3
    }
   ],
   "metadata": {}
  }
 ]
}